File: doo/connect_devices.py
#!/usr/bin/env python3

import yaml
import subprocess
import sys
import os
import platform

# Correct path to your YAML file
DEVICE_FILE = "/home/labuser/sp-1/doo/configs.3/devices.yaml"

def load_devices(file_path):
    with open(file_path, "r") as f:
        data = yaml.safe_load(f)
        if isinstance(data, dict) and "devices" in data:
            return data["devices"]
        return data  # assume it's already a list

def ping_device(ip):
    count = "1"
    cmd = ["ping", "-c", count, ip] if platform.system() != "Windows" else ["ping", "-n", count, ip]
    result = subprocess.run(cmd, stdout=subprocess.DEVNULL)
    return result.returncode == 0

def ssh_to_device(device):
    expect_script = f"""
    set timeout 10
    spawn ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null {device['username']}@{device['hostname']}
    expect {{
        "*assword:" {{
            send "{device['password']}\\r"
        }}
    }}
    interact
    """
    temp_script = "/tmp/ssh_expect_script.exp"
    with open(temp_script, "w") as f:
        f.write(expect_script)

    os.system(f"gnome-terminal -- bash -c 'expect {temp_script}; exec bash'")


def telnet_to_device(device):
    telnet_cmd = f"gnome-terminal -- bash -c 'telnet {device['hostname']}; exec bash'"
    os.system(telnet_cmd)

def connect_device(hostname, devices):
    for device in devices:
        if device["name"].lower() == hostname.lower():
            ip = device["hostname"]
            print(f"==> Connecting to {device['name']} ({ip}) via {device['device_type'].upper()}")
            if not ping_device(ip):
                print(f"✗ {device['name']} is NOT reachable.")
                return
            print(f"✓ {device['name']} is reachable.")
            if "xr" in device["device_type"].lower():
                ssh_to_device(device)
            else:
                telnet_to_device(device)
            return
    print(f"✗ Device '{hostname}' not found in devices.yaml.")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 connect_devices.py <device-name>")
        sys.exit(1)

    hostname = sys.argv[1]
    devices = load_devices(DEVICE_FILE)
    connect_device(hostname, devices)

File: doo/configs.5/connect_devices.pl
#!/usr/bin/env perl
use strict;
use warnings;
use YAML::XS 'LoadFile';
use Net::OpenSSH;
use Net::Telnet;

my $yaml_file = '~/sp-1/doo/configs.3/devices.yaml';  # Ensure this path is correct
my $data = LoadFile($yaml_file);

my @reachable;
my @unreachable;

print "\n--- CE Device Connection Script (Ping + SSH/Telnet) ---\n";

foreach my $device (@{$data->{devices}}) {
    my $name = $device->{name};
    my $ip   = $device->{hostname};
    my $user = $device->{username} || 'cisco';
    my $pass = $device->{password} || 'cisco';
    my $type = $device->{device_type};

    # Only handle CE devices
    next unless $name =~ /-CE-/i;

    print "\n[$name] Pinging $ip ... ";
    my $ping_result = system("ping -c 1 -W 2 $ip > /dev/null 2>&1");

    if ($ping_result == 0) {
        print "Reachable ✅\n";
        push @reachable, $name;

        if ($type eq 'cisco_xr') {
            print "[$name] Connecting via SSH...\n";
            my $ssh = Net::OpenSSH->new($ip, user => $user, password => $pass, timeout => 5);
            if ($ssh->error) {
                warn "SSH failed for $name: " . $ssh->error . "\n";
            } else {
                print $ssh->capture("show version | include IOS");
            }

        } elsif ($type eq 'cisco_ios') {
            print "[$name] Connecting via Telnet...\n";
            my $telnet = Net::Telnet->new(Host => $ip, Timeout => 5, Errmode => 'return');
            $telnet->open($ip);
            $telnet->waitfor('/Username:/i');
            $telnet->print($user);
            $telnet->waitfor('/Password:/i');
            $telnet->print($pass);
            my @output = $telnet->cmd('show version | include IOS');
            print @output;
            $telnet->close;
        } else {
            warn "Unknown device type for $name\n";
        }

    } else {
        print "Unreachable ❌\n";
        push @unreachable, $name;
    }
}

print "\n--- Summary ---\n";
print "Reachable CE Devices:\n  ", join("\n  ", @reachable), "\n" if @reachable;
print "Unreachable CE Devices:\n  ", join("\n  ", @unreachable), "\n" if @unreachable;
